api/agent/call.go

```diff
@@ -368,28 +338,82 @@ func (a *agent) GetCall(opts ...CallOpt) (Call, error) {
 	return &c, nil
 }
 
-func setupCtx(c *call) {
-	ctx, _ := common.LoggerWithFields(c.req.Context(),
-		logrus.Fields{"id": c.ID, "app_id": c.AppID, "route": c.Path})
-	c.req = c.req.WithContext(ctx)
+// setCallPayload sets the GetBody method on the call, if one doesn't exist
+func (c *call) setCallPayload(ctx context.Context, input io.Reader, getBody func() (io.ReadCloser, error)) error {
+	if getBody != nil {
+		c.getBody = getBody
+		return nil
+	}
+
+	// if it has a string method, shortcut that (say, a bytes buffer...)
+	// TODO we could check Bytes() as well but this whole thing should die anyway, hedging...
+	if stringer, ok := input.(interface{ String() string }); ok {
+		c.getBody = func() (io.ReadCloser, error) {
+			return ioutil.NopCloser(strings.NewReader(stringer.String())), nil
+		}
+		return nil
+	}
+
+	// only allow one read of the buffer ever. everybody else must wait. this is
+	// entirely fucked up and can deadlock all succeeding threads for this body,
+	// but go ahead and try just reading the reader into a buffer and see if life
+	// is still worth living. TODO this leaves a situation where if the first reader
+	// doesn't finish then nobody else gets the body ever. NO SOUP FOR YOU
```

Comment:
I think this is fine, all of such I/O reads would typically be done in a go-routine, where we have to assume that first-reader would close/abort properly unlocking the followers as well.

Another weird case here is when a non-eof error happens during ReadFrom(), the followers will block.
We might want to expose ReadFrom() error code for the followers are well.

Reply:
> But when a non-eof error happens during ReadFrom(), with early return
the followers will block. We might want to close to channel and
expose ReadFrom() error code for the followers are well.

good idea

Reply:
done, but the comment stays... there's still a case where the first request could get cut short that the second request could in theory pick up and finish flushing to the cache, that's just where we get into race territory...